### ISSUE TITLE
"Failed to pack replicate object" error

### ISSUE THREAD

**Issue by colin-nolan** [wtsi-npg/baton#140]
As discussed in person earlier, I am encountering a problem where baton is raising a "Failed to pack replicate object" error when used to get information about replicas (using `--replicates`). This error appears in the setup I am using if any of the replica's checksums are not stored on iRODS. 

[iRODS v3.3.1 iCAT server](https://hub.docker.com/r/agaveapi/irods/) running in Docker container:
`docker run -d -p 1247:1247 agaveapi/irods:3.3.1`
```
irodsUserName rods
irodsHost <your_docker_ip>
irodsPort 1247
irodsZone iplant
# Password: rods
```

MWE (ran in [dockerised build of baton](https://github.com/wtsi-hgi/docker-baton) from wtsi-npg master branch):
```
root@5d09b5ae5dde:/scripts# iput test
root@5d09b5ae5dde:/scripts# ils -L
/iplant/home/rods:
  rods              0 demoResc                        0 2016-01-19.13:13 & test
            generic    /data/home/rods/test
root@5d09b5ae5dde:/scripts# jq -n '{"collection": "/iplant/home/rods", "data_object": "test"}' | baton-list --replicate --debug | jq "."
2016-01-19T13:13:45 DEBUG Preparing a query to select 2 columns.
2016-01-19T13:13:45 DEBUG Adding condition 1 of 2: COLL_NAME = /iplant/home/rods.
2016-01-19T13:13:45 DEBUG Made string 0 of 2: op: = value: /iplant/home/rods, len 17, total len 22 [= '/iplant/home/rods'].
2016-01-19T13:13:45 DEBUG Adding condition 1 of 2: DATA_NAME = test.
2016-01-19T13:13:45 DEBUG Made string 1 of 2: op: = value: test, len 4, total len 9 [= 'test'].
2016-01-19T13:13:45 DEBUG Adding condition 1 of 1: DATA_REPL_STATUS = 1.
2016-01-19T13:13:45 DEBUG Made string 0 of 1: op: = value: 1, len 1, total len 6 [= '1'].
2016-01-19T13:13:45 DEBUG Running query ....
2016-01-19T13:13:45 DEBUG Attempting to get chunk 0 of query.
2016-01-19T13:13:45 DEBUG Successfully fetched chunk 0 of query.
2016-01-19T13:13:45 DEBUG Converting 1 rows of results to JSON.
2016-01-19T13:13:45 DEBUG Converting row 0 of 1 to JSON.
2016-01-19T13:13:45 DEBUG Encoding column 0 'collection' value '/iplant/home/rods' as JSON.
2016-01-19T13:13:45 DEBUG Encoding column 1 'data_object' value 'test' as JSON.
2016-01-19T13:13:45 DEBUG Obtained a total of 1 JSON results in 1 chunks.
2016-01-19T13:13:45 DEBUG Preparing a query to select 5 columns.
2016-01-19T13:13:45 DEBUG Adding condition 1 of 2: COLL_NAME = /iplant/home/rods.
2016-01-19T13:13:45 DEBUG Made string 0 of 2: op: = value: /iplant/home/rods, len 17, total len 22 [= '/iplant/home/rods'].
2016-01-19T13:13:45 DEBUG Adding condition 1 of 2: DATA_NAME = test.
2016-01-19T13:13:45 DEBUG Made string 1 of 2: op: = value: test, len 4, total len 9 [= 'test'].
2016-01-19T13:13:45 DEBUG Using zone hint '/iplant/home/rods/test'.
2016-01-19T13:13:45 DEBUG Running query ....
2016-01-19T13:13:45 DEBUG Attempting to get chunk 0 of query.
2016-01-19T13:13:45 DEBUG Successfully fetched chunk 0 of query.
2016-01-19T13:13:45 DEBUG Converting 1 rows of results to JSON.
2016-01-19T13:13:45 DEBUG Converting row 0 of 1 to JSON.
2016-01-19T13:13:45 DEBUG Encoding column 0 'valid' value '1' as JSON.
2016-01-19T13:13:45 DEBUG Encoding column 1 'number' value '0' as JSON.
2016-01-19T13:13:45 DEBUG Encoding column 2 'checksum' value '' as JSON.
2016-01-19T13:13:45 DEBUG Encoding column 3 'resource' value 'demoResc' as JSON.
2016-01-19T13:13:45 DEBUG Encoding column 4 'location' value 'docker.example.com' as JSON.
2016-01-19T13:13:45 DEBUG Obtained a total of 1 JSON results in 1 chunks.
2016-01-19T13:13:45 ERROR Failed to list replicates of '/iplant/home/rods/test': error -1 Failed to pack replicate object.
2016-01-19T13:13:45 ERROR Failed to pack replicate object.
2016-01-19T13:13:45 DEBUG Processed 1 items with 1 errors.
{
  "collection": "/iplant/home/rods",
  "data_object": "test",
  "error": {
    "message": "Failed to pack replicate object",
    "code": -1
  }
}
root@5d09b5ae5dde:/scripts# ichksum test
    test                              d41d8cd98f00b204e9800998ecf8427e
Total checksum performed = 1, Failed checksum = 0
root@5d09b5ae5dde:/scripts# ils -L
/iplant/home/rods:
  rods              0 demoResc                        0 2016-01-19.13:13 & test
    d41d8cd98f00b204e9800998ecf8427e        generic    /data/home/rods/test
root@5d09b5ae5dde:/scripts# jq -n '{"collection": "/iplant/home/rods", "data_object": "test"}' | baton-list --replicate --debug | jq "."
2016-01-19T13:14:06 DEBUG Preparing a query to select 2 columns.
2016-01-19T13:14:06 DEBUG Adding condition 1 of 2: COLL_NAME = /iplant/home/rods.
2016-01-19T13:14:06 DEBUG Made string 0 of 2: op: = value: /iplant/home/rods, len 17, total len 22 [= '/iplant/home/rods'].
2016-01-19T13:14:06 DEBUG Adding condition 1 of 2: DATA_NAME = test.
2016-01-19T13:14:06 DEBUG Made string 1 of 2: op: = value: test, len 4, total len 9 [= 'test'].
2016-01-19T13:14:06 DEBUG Adding condition 1 of 1: DATA_REPL_STATUS = 1.
2016-01-19T13:14:06 DEBUG Made string 0 of 1: op: = value: 1, len 1, total len 6 [= '1'].
2016-01-19T13:14:06 DEBUG Running query ....
2016-01-19T13:14:06 DEBUG Attempting to get chunk 0 of query.
2016-01-19T13:14:06 DEBUG Successfully fetched chunk 0 of query.
2016-01-19T13:14:06 DEBUG Converting 1 rows of results to JSON.
2016-01-19T13:14:06 DEBUG Converting row 0 of 1 to JSON.
2016-01-19T13:14:06 DEBUG Encoding column 0 'collection' value '/iplant/home/rods' as JSON.
2016-01-19T13:14:06 DEBUG Encoding column 1 'data_object' value 'test' as JSON.
2016-01-19T13:14:06 DEBUG Obtained a total of 1 JSON results in 1 chunks.
2016-01-19T13:14:06 DEBUG Preparing a query to select 5 columns.
2016-01-19T13:14:06 DEBUG Adding condition 1 of 2: COLL_NAME = /iplant/home/rods.
2016-01-19T13:14:06 DEBUG Made string 0 of 2: op: = value: /iplant/home/rods, len 17, total len 22 [= '/iplant/home/rods'].
2016-01-19T13:14:06 DEBUG Adding condition 1 of 2: DATA_NAME = test.
2016-01-19T13:14:06 DEBUG Made string 1 of 2: op: = value: test, len 4, total len 9 [= 'test'].
2016-01-19T13:14:06 DEBUG Using zone hint '/iplant/home/rods/test'.
2016-01-19T13:14:06 DEBUG Running query ....
2016-01-19T13:14:06 DEBUG Attempting to get chunk 0 of query.
2016-01-19T13:14:06 DEBUG Successfully fetched chunk 0 of query.
2016-01-19T13:14:06 DEBUG Converting 1 rows of results to JSON.
2016-01-19T13:14:06 DEBUG Converting row 0 of 1 to JSON.
2016-01-19T13:14:06 DEBUG Encoding column 0 'valid' value '1' as JSON.
2016-01-19T13:14:06 DEBUG Encoding column 1 'number' value '0' as JSON.
2016-01-19T13:14:06 DEBUG Encoding column 2 'checksum' value 'd41d8cd98f00b204e9800998ecf8427e' as JSON.
2016-01-19T13:14:06 DEBUG Encoding column 3 'resource' value 'demoResc' as JSON.
2016-01-19T13:14:06 DEBUG Encoding column 4 'location' value 'docker.example.com' as JSON.
2016-01-19T13:14:06 DEBUG Obtained a total of 1 JSON results in 1 chunks.
2016-01-19T13:14:06 DEBUG Obtained replicates of '/iplant/home/rods/test'.
2016-01-19T13:14:06 DEBUG Processed 1 items with 0 errors.
{
  "collection": "/iplant/home/rods",
  "data_object": "test",
  "replicates": [
    {
      "resource": "demoResc",
      "location": "docker.example.com",
      "checksum": "d41d8cd98f00b204e9800998ecf8427e",
      "valid": true,
      "number": 0
    }
  ]
}
```

However, the same problem is not encountered in the sequencing iRODS server (baton 0.15.0 install on hgi-mercury-farm3 used):
```
cn13@hgi-mercury-farm3:~$ iput test
cn13@hgi-mercury-farm3:~$ ils -L
/Sanger1/home/cn13:
  cn13              0 wtsiusers                       0 2016-01-19.13:12 & test
            generic    /data/home/cn13/test
  C- /Sanger1/home/cn13/myCollection
cn13@hgi-mercury-farm3:~$ jq -n '{"collection": "/Sanger1/home/cn13", "data_object": "test"}' | baton-list --replicate --debug | jq "."
2016-01-19T13:12:50 DEBUG Preparing a query to select 2 columns.
2016-01-19T13:12:50 DEBUG Adding condition 1 of 2: COLL_NAME = /Sanger1/home/cn13.
2016-01-19T13:12:50 DEBUG Made string 0 of 2: op: = value: /Sanger1/home/cn13, len 18, total len 23 [= '/Sanger1/home/cn13'].
2016-01-19T13:12:50 DEBUG Adding condition 1 of 2: DATA_NAME = test.
2016-01-19T13:12:50 DEBUG Made string 1 of 2: op: = value: test, len 4, total len 9 [= 'test'].
2016-01-19T13:12:50 DEBUG Adding condition 1 of 1: DATA_REPL_STATUS = 1.
2016-01-19T13:12:50 DEBUG Made string 0 of 1: op: = value: 1, len 1, total len 6 [= '1'].
2016-01-19T13:12:50 DEBUG Running query ....
2016-01-19T13:12:50 DEBUG Attempting to get chunk 0 of query.
2016-01-19T13:12:50 DEBUG Successfully fetched chunk 0 of query.
2016-01-19T13:12:50 DEBUG Converting 1 rows of results to JSON.
2016-01-19T13:12:50 DEBUG Converting row 0 of 1 to JSON.
2016-01-19T13:12:50 DEBUG Encoding column 0 'collection' value '/Sanger1/home/cn13' as JSON.
2016-01-19T13:12:50 DEBUG Encoding column 1 'data_object' value 'test' as JSON.
2016-01-19T13:12:50 DEBUG Obtained a total of 1 JSON results in 1 chunks.
2016-01-19T13:12:50 DEBUG Preparing a query to select 3 columns.
2016-01-19T13:12:50 DEBUG Adding condition 1 of 2: COLL_NAME = /Sanger1/home/cn13.
2016-01-19T13:12:50 DEBUG Made string 0 of 2: op: = value: /Sanger1/home/cn13, len 18, total len 23 [= '/Sanger1/home/cn13'].
2016-01-19T13:12:50 DEBUG Adding condition 1 of 2: DATA_NAME = test.
2016-01-19T13:12:50 DEBUG Made string 1 of 2: op: = value: test, len 4, total len 9 [= 'test'].
2016-01-19T13:12:50 DEBUG Using zone hint '/Sanger1/home/cn13/test'.
2016-01-19T13:12:50 DEBUG Running query ....
2016-01-19T13:12:50 DEBUG Attempting to get chunk 0 of query.
2016-01-19T13:12:50 DEBUG Successfully fetched chunk 0 of query.
2016-01-19T13:12:50 DEBUG Converting 1 rows of results to JSON.
2016-01-19T13:12:50 DEBUG Converting row 0 of 1 to JSON.
2016-01-19T13:12:50 DEBUG Encoding column 0 'valid' value '1' as JSON.
2016-01-19T13:12:50 DEBUG Encoding column 1 'number' value '0' as JSON.
2016-01-19T13:12:50 DEBUG Encoding column 2 'checksum' value '' as JSON.
2016-01-19T13:12:50 DEBUG Obtained a total of 1 JSON results in 1 chunks.
2016-01-19T13:12:50 DEBUG Obtained replicates of '/Sanger1/home/cn13/test'.
2016-01-19T13:12:50 DEBUG Processed 1 items with 0 errors.
{
  "replicate": [
    {
      "valid": true,
      "number": 0
    }
  ],
  "data_object": "test",
  "collection": "/Sanger1/home/cn13"
}
```

**Comment by keithj** [comment]
Thanks for the report. The error message comes from [packing the replicate information](https://github.com/wtsi-npg/baton/blob/20a895eac40ccbc62148da90f2edacde18de4bfc/src/json.c#L439-L448) as you described. If iRODS can return data in normal use that baton can't
handle, that's a baton bug. I'll look into it.

**Comment by irinaColgiu** [comment]
Hi Keith,

I am getting the same error when trying out baton 0.16 with different files from Sanger1 zone:

ic4@hgi-serapis-farm3:~$ jq -n '{collection: "/Sanger1/home/ic4", data_object: "users.txt"}' | /software/gapi/pkg/baton/0.16.0/bin/baton-list --avu --replicate
2016-03-10T16:58:11 ERROR Failed to list replicates of '/Sanger1/home/ic4/users.txt': error -1 Failed to pack replicate object.
2016-03-10T16:58:11 ERROR Failed to pack replicate object.
{"data_object": "users.txt", "collection": "/Sanger1/home/ic4", "error": {"code": -1, "message": "Failed to pack replicate object"}}

The output with --debug looks almost identical to what Colin has posted above.

The interesting thing is that the issue gets fixed if one runs ichksum. Not sure why.
BR,
Irina

**Comment by keithj** [comment]
The error is caused by the absence of the checksum. `baton` checks for a checksum and if there isn't one, it will complain. The error message could be more specific.

 When I wrote that code, I expected checksums always to be present. In NPG we always trigger a checksum by using `iput -k`.

I think that the current behaviour is incorrect and will fix it as soon as I have time (patches welcome!)

**Comment by keithj** [comment]
We have two options here; either include the `checksum` key with an explicit value of JSON `null`, or omit the `checksum` key entirely from replicates that do not have it. I'm inclined to go with the former because then all replicate JSON will have the same properties (and consumers will not have to test the presence of the `checksum` property) and it makes the missing value explicit.

**Comment by keithj** [comment]
Fixed by #153
